eclair-core/src/main/scala/fr/acinq/eclair/channel/Helpers.scala

```diff
@@ -627,366 +630,394 @@ object Helpers {
+              val htlcRedeemScript = htlcsRedeemScripts(txOut.publicKeyScript)
+              withTxGenerationLog("htlc-penalty") {
+                Transactions.makeHtlcPenaltyTx(commitTx, outputIndex, htlcRedeemScript, localParams.dustLimit, localParams.defaultFinalScriptPubKey, feeratePerKwPenalty).map(htlcPenalty => {
+                  val sig = keyManager.sign(htlcPenalty, keyManager.revocationPoint(channelKeyPath), remotePerCommitmentSecret, TxOwner.Local, commitmentFormat)
+                  Transactions.addSigs(htlcPenalty, sig, remoteRevocationPubkey)
                 })
               }
-            })
-            val revokedCommitPublished1 = revokedCommitPublished.copy(claimHtlcDelayedPenaltyTxs = revokedCommitPublished.claimHtlcDelayedPenaltyTxs ++ penaltyTxs)
-            (revokedCommitPublished1, penaltyTxs)
-          }).getOrElse((revokedCommitPublished, Nil))
-      } else {
-        (revokedCommitPublished, Nil)
+            }.toList.flatten
+
+            RevokedCommitPublished(
+              commitTx = commitTx,
+              claimMainOutputTx = mainTx,
+              mainPenaltyTx = mainPenaltyTx,
+              htlcPenaltyTxs = htlcPenaltyTxs,
+              claimHtlcDelayedPenaltyTxs = Nil, // we will generate and spend those if they publish their HtlcSuccessTx or HtlcTimeoutTx
+              irrevocablySpent = Map.empty
+            )
+          })
+      }
+
+      /**
+       * Claims the output of an [[HtlcSuccessTx]] or [[HtlcTimeoutTx]] transaction using a revocation key.
+       *
+       * In case a revoked commitment with pending HTLCs is published, there are two ways the HTLC outputs can be taken as punishment:
+       * - by spending the corresponding output of the commitment tx, using [[HtlcPenaltyTx]] that we generate as soon as we detect that a revoked commit
+       * as been spent; note that those transactions will compete with [[HtlcSuccessTx]] and [[HtlcTimeoutTx]] published by the counterparty.
+       * - by spending the delayed output of [[HtlcSuccessTx]] and [[HtlcTimeoutTx]] if those get confirmed; because the output of these txs is protected by
+       * an OP_CSV delay, we will have time to spend them with a revocation key. In that case, we generate the spending transactions "on demand",
+       * this is the purpose of this method.
+       *
+       * NB: when anchor outputs is used, htlc transactions can be aggregated in a single transaction if they share the same
+       * lockTime (thanks to the use of sighash_single | sighash_anyonecanpay), so we may need to claim multiple outputs.
+       */
+      def claimHtlcTxOutputs(keyManager: ChannelKeyManager, commitments: Commitments, revokedCommitPublished: RevokedCommitPublished, htlcTx: Transaction, feeEstimator: FeeEstimator)(implicit log: LoggingAdapter): (RevokedCommitPublished, Seq[ClaimHtlcDelayedOutputPenaltyTx]) = {
```

Comment:
This one makes sense to keep plural, because our peer may have batched multiple htlc txs into a single one, so the name is good.

Reply:
Maybe `claimHtlcTxOutputs`->`claimHtlcOutputs` though?